### PR TITLE
Implement low-level git diff functionality.

### LINKED ISSUE #10848: Implement low-level diff functionality
Part of #10323 

* Implement low-level git diff functionality
* Make sure initial commits are covered
* Add ability to find renames or copies

### PR DESCRIPTION
**Pull request by EnTeQuAk**
This implements all the necessary changes to diff two versions against
each other and also handles initial commits.

Fixes #10848

### CODE REVIEW

File: src/olympia/lib/git.py [hunk 1]
```diff
@@ -370,3 +393,91 @@ def iter_tree(self, tree):
...
+            context_lines=sys.maxsize,
+            interhunk_lines=0)
+
+        opts = pygit2.GIT_DIFF_FIND_RENAMES | pygit2.GIT_DIFF_FIND_COPIES
+        diff.find_similar(opts, SIMILARITY_THRESHOLD, SIMILARITY_THRESHOLD)
+
+        # TODO (cgrebs): double-check if we are indeed processing files
+        # multiple times (e.g on renames, copies etc)
+        # I think this happens when renames are < similarity threshold or so
+        # needs some more tests though...
```
Comment (eviljeff) [inline src/olympia/lib/git.py]: Issue for follow-up?

File: src/olympia/lib/git.py [hunk 2]
```diff
@@ -346,6 +366,9 @@ def get_root_tree(self, commit):
 
         This doesn't contain the ``EXTRACTED_PREFIX`` prefix folder.
         """
+        if isinstance(commit, six.string_types):
+            commit = self.git_repository[commit]
```
Comment (eviljeff) [inline src/olympia/lib/git.py]: When does this happen?

**Comment by EnTeQuAk** [comment]
> Is there a test that checks SIMILARITY_THRESHOLD and file moves? (Might have missed it - some of the tests are pretty long)

Sorry, I forgot to mention this. I'll add this in a separate PR via https://github.com/mozilla/addons-server/issues/10849 since renames and copies will involve even more test-helper code and I wanted this to be easier to review here.

File: src/olympia/lib/git.py (hunk 2 above)
Comment (EnTeQuAk) [reply src/olympia/lib/git.py]: Added a comment in 58e3ce9ba2, it basically happens in the drf api serializers which simply grab `version.git_hash` and pass that on as `commit`

When that function is used internally (to hide the `extracted/` folder implementation detail) we are passing along actual commit objects.

File: src/olympia/lib/git.py (hunk 1 above)
Comment (EnTeQuAk) [reply src/olympia/lib/git.py]: I'll do that as part of #10849

**Pull request merged**